walk/git.go

```diff
@@ -6,40 +6,54 @@ import (
 	"io/fs"
 	"os"
 	"path/filepath"
+	"time"
+
+	"github.com/go-git/go-git/v5/plumbing/format/index"
 
 	"github.com/charmbracelet/log"
 
 	"github.com/go-git/go-git/v5"
 )
 
 type gitWalker struct {
-	root  string
-	paths chan string
-	repo  *git.Repository
+	root          string
+	paths         chan string
+	repo          *git.Repository
+	allFiles      bool
+	relPathOffset int
 }
 
-func (g *gitWalker) Root() string {
+func (g gitWalker) Root() string {
 	return g.root
 }
 
-func (g *gitWalker) Walk(ctx context.Context, fn WalkFunc) error {
-	// for quick relative paths
-	relPathOffset := len(g.root) + 1
-
-	relPathFn := func(path string) (relPath string) {
-		if len(path) >= relPathOffset {
-			relPath = path[relPathOffset:]
-		}
-		return
+func (g gitWalker) relPath(path string) (string, error) {
+	// quick optimization for the majority of use cases
+	if len(path) >= g.relPathOffset && path[:len(g.root)] == g.root {
+		return path[g.relPathOffset:], nil
 	}
+	// fallback to proper relative path resolution
+	return filepath.Rel(g.root, path)
+}
 
+func (g gitWalker) Walk(ctx context.Context, fn WalkFunc) error {
 	idx, err := g.repo.Storer.Index()
 	if err != nil {
 		return fmt.Errorf("failed to open git index: %w", err)
 	}
 
 	// cache in-memory whether a path is present in the git index
-	var cache map[string]bool
+	var cache map[string]*index.Entry
+
+	hasChanged := func(entry *index.Entry, info os.FileInfo) bool {
+		if g.allFiles {
+			// user wants all files regardless of changes
+			return true
+		}
+		// we only want to emit files that have changed when compared with the index
+		// mod time comparison is done with EPOCH (second) precision as per the POSIX spec
+		return entry.ModifiedAt.Truncate(time.Second) != info.ModTime().Truncate(time.Second)
```

Comment:
Is the entry.ModifiedAt coming from the git commit?

Reply:
Yup